Set 'calculating bill run' status when generating the bill run summary

StuAA78 opened the pull request:
https://trello.com/c/weKYn1nO/1833-325-defect-view-bill-run-endpoint-times-out-whilst-create-transaction-end-point-is-running-and-shortly-after-completed

This defect was raised due to timeout errors when adding large numbers of transactions to a bill run, which seems to be caused by the bill run summary taking too long to calculate. A solution we're exploring is to set the bill run status to `calculating_summary` (or similar) when the summary is being calculated.

This change was originally intended to handle just the initial step of setting the status while the summary is being generated. However while working on it I quickly realised this wasn't the right way to go about it.

Instead this change implements the feature as follows:
* If the summary exists for the bill run then this is returned (as is currently the case)
* If the summary doesn't exist then the bill run is generated
* If the bill run hasn't been generated within a set time (defined in `.env` as `BILL_RUN_SUMMARY_TIMEOUT`) then a holding response is returned

StuAA78 commented:
Failing Travis CI tests were due to the `BILL_RUN_SUMMARY_TIMEOUT` environment variable being missing from `.travis.yml` -- once I realised I needed to add it, the tests started passing.

Cruikshanks commented on app/services/view_bill_run.js:
```diff
@@ -12,12 +13,36 @@ async function call (request) {
-  if (!billRun.summary_data) {
-    await GenerateBillRunSummary.call(request.regime, billRun)
+  // If the bill run already has a summary then return it
+  if (billRun.summary_data) {
```
So having discussed this with @StuAA78 the issue we have is that

- we determine a bill run summary needs generating by looking at the value in the bill run record
- we only update those values when the summary is calculated
- for large bill runs that can be a long time (WRLS reported timeouts after 30 secs so assume at least as long as that)

So the scenario we could face is

- request received
- no summary found
- bill summary triggered but will take 30 secs to complete
- we respond with 'holding response' after 3 seconds
- client waits 5 secs and then tries again
- request received
- no summary found
- another bill summary triggered that will take 30 secs to complete
- we respond with 'holding response' after 3 seconds
- client waits 5 secs and then tries again
- [repeat until first bill run summary completes]

We can't control how clients interact with the API so this is a feasible scenario. We also don't know what impact running multiple bill summaries will have on the service.

So I think we need to alter the status of the bill run record such that we can see if a summary is already in the process of being generated. If it is we can immediately return the holding response and prevent starting another process. This should nullify the problem for now till such time as we have thought again about how bill run summaries are triggered, and why they have to be such a performance hog.

Cruikshanks reviewed (changes requested):
Implementation is great and I think it's a nifty solution to the situation we find ourselves in.

But I would want to protect against the scenario I've outlined before the merge the change.

So implementing the change as discussed and then we should be good to go!

StuAA78 commented:
We now write the `generating_summary` status to the database when it's being generated and check for this when to ensure we don't set up multiple threads all generating the summary.

There's one other place where the summary is generated and could potentially time out: when a bill run is sent without a summary. I've implemented the fix here as well. Note that once the summary is generated it doesn't actually send the bill run until the endpoint is hit again -- if we sent it as soon as the summary is generated then the end user wouldn't get confirmation that the bill run was successfully sent (as sending a bill run that's already sent returns an error). If this becomes an issue we can rework the send bill run process, however as this scenario hasn't yet been raised as an issue I don't think it's worth spending more time on it apart from this pre-emptive fix.

To explain why I removed one of the unit tests: I realised that the scenario it covered (a bill run is sent with no transactions) can't happen as a bill run can only be approved for sending if it contains a transaction, and once approved it can't be amended to remove the transaction.

Cruikshanks reviewed (approved):
Looks good to me 👍
